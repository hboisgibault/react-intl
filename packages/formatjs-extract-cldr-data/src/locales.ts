/*
 * Copyright 2015, Yahoo Inc.
 * Copyrights licensed under the New BSD License.
 * See the accompanying LICENSE file for terms.
 */
// These are the exceptions to the default algorithm for determining a locale's
// parent locale.
import * as AVAILABLE_LOCALES from 'cldr-core/availableLocales.json';

export function getAllLocales() {
  return AVAILABLE_LOCALES.availableLocales.full.sort();
}

export function getAllLanguages() {
  return new Set(
    AVAILABLE_LOCALES.availableLocales.full.map(l => l.split('-')[0])
  );
}
