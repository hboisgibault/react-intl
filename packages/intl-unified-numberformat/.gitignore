src/en.ts
src/locales.ts
src/units-constants.ts
src/currency-digits.json
src/ilnd-numbers.json