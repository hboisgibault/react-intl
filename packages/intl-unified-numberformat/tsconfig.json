{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "lib": ["dom", "es5", "esnext.intl", "es2017.intl", "es2018.intl"],
    "outDir": "dist",
    "rootDir": "src"
  },
  "include": ["src", "src/**/*.json"],
  "references": [
    {"path": "../intl-pluralrules"},
    {"path": "../formatjs-extract-cldr-data"},
    {"path": "../intl-utils"}
  ]
}
