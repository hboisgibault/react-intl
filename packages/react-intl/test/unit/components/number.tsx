import * as React from 'react';
import {mount} from 'enzyme';
import {FormattedNumber, FormattedNumberParts} from '../../../src';
import {createIntl} from '../../../src/components/provider';
import {mountFormattedComponentWithProvider} from '../testUtils';
import {UnifiedNumberFormatOptions} from '@formatjs/intl-unified-numberformat';
const mountWithProvider = mountFormattedComponentWithProvider(FormattedNumber);
const mountPartsWithProvider = mountFormattedComponentWithProvider(
  FormattedNumberParts
);

describe('<FormattedNumber>', () => {
  let intl;
  const onError = jest.fn();
  beforeEach(() => {
    onError.mockClear();
    intl = createIntl({
      locale: 'en',
      onError,
    });
  });

  it('has a `displayName`', () => {
    expect(typeof FormattedNumber.displayName).toBe('string');
  });

  it('throws when <IntlProvider> is missing from ancestry', () => {
    expect(() => mount(<FormattedNumber value={0} />)).toThrow(
      '[React Intl] Could not find required `intl` object. <IntlProvider> needs to exist in the component ancestry.'
    );
  });

  it('renders "NaN" in a <span> when no `value` prop is provided', () => {
    const rendered = mountWithProvider({value: undefined}, intl);

    expect(rendered.text()).toBe('NaN');
  });

  it('renders a formatted number in a <>', () => {
    const num = 1000;

    const rendered = mountWithProvider({value: num}, intl);

    expect(rendered.text()).toBe(intl.formatNumber(num));
  });

  it('renders a formatted number w/o textComponent', () => {
    const num = 1000;

    const rendered = mountWithProvider(
      {value: num},
      {...intl, textComponent: null}
    );

    expect(rendered.text()).toBe(intl.formatNumber(num));
  });

  it('accepts valid Intl.NumberFormat options as props', () => {
    const num = 0.5;
    const options = {style: 'percent' as UnifiedNumberFormatOptions['style']};

    const rendered = mountWithProvider({value: num, ...options}, intl);

    expect(rendered.text()).toBe(intl.formatNumber(num, options));
  });

  it('falls back and warns on invalid Intl.NumberFormat options', () => {
    const rendered = mountWithProvider(
      {value: 0, style: 'invalid' as any},
      intl
    );

    expect(rendered.text()).toBe('0');
    expect(onError.mock.calls[0][0].code).toMatchSnapshot();
    expect(onError).toHaveBeenCalledTimes(1);
  });

  it('accepts `format` prop', () => {
    intl = createIntl({
      locale: 'en',
      formats: {
        number: {
          percent: {
            style: 'percent',
            minimumFractionDigits: 2,
          },
        },
      },
    });

    const num = 0.505;
    const format = 'percent';

    const rendered = mountWithProvider({value: num, format}, intl);

    expect(rendered.text()).toBe(intl.formatNumber(num, {format}));
  });

  it('supports function-as-child pattern', () => {
    const num = Date.now();

    const spy = jest.fn().mockImplementation(() => <span>Jest</span>);
    const rendered = mountWithProvider({value: num, children: spy}, intl).find(
      'span'
    );

    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy.mock.calls[0]).toEqual([intl.formatNumber(num)]);

    expect(rendered.type()).toBe('span');
    expect(rendered.text()).toBe('Jest');
  });
});

function NOOP(_: Intl.NumberFormatPart[]) {
  return null;
}

describe('<FormattedNumberParts>', function() {
  let intl;
  const children = jest.fn();

  beforeEach(() => {
    console.error = jest.fn();
    intl = createIntl({
      locale: 'en',
    });
    children.mockClear();
  });

  it('has a `displayName`', () => {
    expect(typeof FormattedNumberParts.displayName).toBe('string');
  });

  it('throws when <IntlProvider> is missing from ancestry', () => {
    expect(() =>
      mount(<FormattedNumberParts value={0} children={NOOP} />)
    ).toThrow(
      '[React Intl] Could not find required `intl` object. <IntlProvider> needs to exist in the component ancestry.'
    );
  });

  it('accepts valid Intl.NumberFormat options as props', () => {
    const num = 0.5;
    const options = {
      style: 'percent' as UnifiedNumberFormatOptions['style'],
      children,
    };

    mountPartsWithProvider({value: num, ...options}, intl);

    expect(children.mock.calls[0][0]).toEqual(
      intl.formatNumberToParts(num, options)
    );
  });

  it('accepts `format` prop', () => {
    intl = createIntl({
      locale: 'en',
      formats: {
        number: {
          percent: {
            style: 'percent',
            minimumFractionDigits: 2,
          },
        },
      },
    });

    const num = 0.505;
    const format = 'percent';

    mountPartsWithProvider({value: num, format, children}, intl);
    expect(children.mock.calls[0][0]).toEqual(
      intl.formatNumberToParts(num, {format})
    );
  });
});
