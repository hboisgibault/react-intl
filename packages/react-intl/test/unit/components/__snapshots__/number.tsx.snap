// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormattedNumber> falls back and warns on invalid Intl.NumberFormat options 1`] = `"FORMAT_ERROR"`;
